interfaces/prompting/requestrules/requestrules.go

```diff
@@ -0,0 +1,877 @@
+		changeOccurred = false
+		return nil, err
+	}
+	newRule.ID = origRule.ID
+	err, conflicts, conflictingPermission := rdb.addRuleToTree(newRule)
+	if err != nil {
+		rdb.addRuleToTree(origRule) // ignore any new error
+		// origRule was successfully removed before, so it should now be able
+		// to be successfully re-added without error, and all is unchanged.
+		changeOccurred = false
+		return nil, fmt.Errorf("%w: conflicts: %+v, Permission: '%s'", err, conflicts, conflictingPermission)
+	}
+	rdb.removeRuleWithID(origRule.ID) // no error can occur, we just confirmed the rule exists
+	if err := rdb.addRule(newRule); err != nil {
+		// Should not occur
+		return nil, fmt.Errorf("internal error: %v", err)
+	}
+	changeOccurred = true
+	return newRule, nil
+}
+
+// Rules returns all rules which apply to the given user.
+func (rdb *RuleDB) Rules(user uint32) []*Rule {
+	rdb.mutex.Lock()
+	defer rdb.mutex.Unlock()
+	ruleFilter := func(rule *Rule) bool {
+		return rule.User == user
+	}
+	return rdb.rulesInternal(ruleFilter)
+}
+
+// rulesInternal returns all rules matching the given filter.
+//
+// TODO: store rules separately per user, snap, and interface, so actions which
+// look up or delete all rules for a given user/snap/interface are much faster.
+// This is safe, since rules must each apply to exactly one user, snap and
+// interface, but may apply to multiple permissions.
+func (rdb *RuleDB) rulesInternal(ruleFilter func(rule *Rule) bool) []*Rule {
+	rules := make([]*Rule, 0)
+	currTime := time.Now()
+	needToSave := false
+	defer func() {
+		if needToSave {
+			rdb.save()
+		}
+	}()
+	for _, rule := range rdb.rules {
+		expired, err := rule.Expired(currTime)
+		if err != nil {
+			// Issue with expiration, this should not occur
+		}
+		if expired {
+			needToSave = true
+			rdb.removeRuleFromTree(rule)
+			rdb.removeRuleWithID(rule.ID)
+			data := map[string]string{"removed": "expired"}
+			rdb.notifyRule(rule.User, rule.ID, data)
+			continue
+		}
+		if ruleFilter(rule) {
+			rules = append(rules, rule)
+		}
+	}
+	return rules
+}
+
+// RulesForSnap returns all rules which apply to the given user and snap.
+func (rdb *RuleDB) RulesForSnap(user uint32, snap string) []*Rule {
+	rdb.mutex.Lock()
+	defer rdb.mutex.Unlock()
+	ruleFilter := func(rule *Rule) bool {
+		return rule.User == user && rule.Snap == snap
+	}
+	return rdb.rulesInternal(ruleFilter)
+}
+
+// RulesForInterface returns all rules which apply to the given user and
+// interface.
+func (rdb *RuleDB) RulesForInterface(user uint32, iface string) []*Rule {
+	rdb.mutex.Lock()
+	defer rdb.mutex.Unlock()
+	ruleFilter := func(rule *Rule) bool {
+		return rule.User == user && rule.Interface == iface
+	}
+	return rdb.rulesInternal(ruleFilter)
+}
+
+// RulesForSnapInterface returns all rules which apply to the given user, snap,
+// and interface.
+func (rdb *RuleDB) RulesForSnapInterface(user uint32, snap string, iface string) []*Rule {
```

Comment:
this is not tested

Reply:
Added better coverage for now. When we support more than just the "home" interface, tests can be much more interesting.